mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+            # NOTE - mtai @ davidmarin - we THEN iterate through the loop and continue on
+
+            # Build each step
+            hadoop_job = self._build_dataproc_hadoop_job(step_num)
+
+            # Clean-up step name
+            step_name = '%s - Step %d of %d' % (self._job_key, step_num + 1, self._num_steps())
+            cleansed_step_name = _cleanse_gcp_job_id(step_name)
+
+            # Submit it
+            log.info('Submitting Dataproc Hadoop Job - %s', cleansed_step_name)
+            job_id = self._submit_hadoop_job(cleansed_step_name, hadoop_job)
+            log.info('Submitted Dataproc Hadoop Job - %s', job_id)
+
+            assert job_id == cleansed_step_name
+
+            # Wait for Job completion
+            self._wait_for_step_to_complete(job_id, step_num=step_num, num_steps=total_steps)
+
+            log.info('Completed Dataproc Hadoop Job - %s', job_id)
+
+    def _wait_for_step_to_complete(self, job_id, step_num=None, num_steps=None):
+        """Helper for _wait_for_step_to_complete(). Wait for
+        step with the given ID to complete, and fetch counters.
+        If it fails, attempt to diagnose the error, and raise an
+        exception.
+
+        This also adds an item to self._log_interpretations
+        """
+        log_interpretation = dict(job_id=job_id)
+        self._log_interpretations.append(log_interpretation)
+
+        while True:
+            # don't antagonize Dataproc's throttling
+            log.debug('Waiting %.1f seconds...' %
+                      self._opts['check_api_status_every'])
+            time.sleep(self._opts['check_api_status_every'])
+
+            # https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#JobStatus
+            job_result = self._get_job(job_id)
+
+            job_start_time = job_result['status']['stateStartTime']
+            job_state = job_result['status']['state']
+
+            log.info('%s  %s - %s' % (job_id, job_start_time, job_state))
+
+            # TODO - mtai @ davidmarin - State checking is NOT tested at this point, see included URL for Dataproc Job states
+            # NOTE -  mtai @ davidmarin - https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.jobs#State
+            if job_state not in ('DONE', 'ERROR', 'CANCELLED'):
+                continue
+
+            # we're done, will return at the end of this
+            if job_state == 'DONE':
+                break
+
+            elif job_state == 'ERROR':
+                raise StepFailedException(step_num=step_num, num_steps=num_steps)
+
+    def _step_input_uris(self, step_num):
+        """Get the gs:// URIs for input for the given step."""
+        if step_num == 0:
+            return [self._upload_mgr.uri(path)
+                    for path in self._get_input_paths()]
+        else:
+            # put intermediate data in HDFS
+            return ['hdfs:///tmp/mrjob/%s/step-output/%s/' % (
```

Comment:
You just copied this from EMR, but realized that we should use `%d` for the step num.

Thinking I'd like to standardize on `'%05d' % step_num` (0-indexed). This is more of a note for myself.